### PR TITLE
enable rebaseOptIn by the governor

### PR DESCRIPTION
**Pull request by sparrowDom**
- enable the governor to toggle rebaseOptIn on
- add events to rebase opt in/out
- remove copy artefacts that is no longer needed
- enable rebasing for [eigenLayer contract](https://etherscan.io/address/0xa4c637e0f704745d182e4d38cab7e7485321d059)

### CODE REVIEW

**Review by DanielVF — DISMISSED** [review]
The only change is that the events should fire on accounting changes, not user intents. In this way, in the future, we can completely offline calculate a users balances. (once we have calculated a one time snapshot of the old events we are missing.)

All this really means in practice is that we need to also emit an event in the account migration internal function, if we are making any changes there. We also might want to slight change these event names to something that does not contain "opt" since sometimes it would be user action, and sometimes automatic.


image

```python
import boa
from contextlib import contextmanager

oeth = boa.load_abi("brownie/abi/oeth.json", name="OETH").at("0x856c4Efb76C1D1AE02e20CEB03A2A6a08b0b8dC3")
dripper = boa.load_abi("brownie/abi/dripper.json", name="Dripper").at("0xc0F42F73b8f01849a2DD99753524d4ba14317EB3")

EIG = "0xa4C637e0F704745D182e4D38cAb7E7485321d059"
UNISWAP_POOL = '0x52299416C469843F4e0d54688099966a6c7d720f'


def stats():
    return {
        'totalSupply': oeth.totalSupply(),
        'nonRebasingSupply': oeth.nonRebasingSupply(),
        'rebasingCredits': oeth.rebasingCredits(),
        'rebasingCreditsPerTokenHighres': oeth.rebasingCreditsPerTokenHighres(),
        'EIG_rebaseState': oeth.rebaseState(EIG),
        'EIG_balanceOf': oeth.balanceOf(EIG),
        'EIG_credits': oeth.creditsBalanceOfHighres(EIG)[0],
        'EIG_creditsPerToken': oeth.creditsBalanceOfHighres(EIG)[1],
        'UNISWAP_rebaseState': oeth.rebaseState(UNISWAP_POOL),
        'UNISWAP_balanceOf': oeth.balanceOf(UNISWAP_POOL),
        'UNISWAP_credits': oeth.creditsBalanceOfHighres(UNISWAP_POOL)[0],
        'UNISWAP_creditsPerToken': oeth.creditsBalanceOfHighres(UNISWAP_POOL)[1],
    }

@contextmanager
def with_stats(name):
    print("----"+name+"----")
    before = stats()
    yield
    after = stats()
    for k in before.keys():
        diff = after[k] - before[k]
        endcolor = '\x1b[0m'
        if diff > 0:
            color = '\x1b[0;32;40m'
        elif diff < 0:
            color = '\x1b[0;33;41m'
        else:
            color = endcolor
        print(color, k, diff / 10**18, diff, '|', before[k], '->', after[k], endcolor)


boa.env.fork(url="http://127.0.0.1:8545/")
print("Secenario 1")
with boa.env.anchor():
    with with_stats("Rebase after governance"):
        dripper.collectAndRebase()


print("")
print("Secenario 2")
with boa.env.anchor():
    with with_stats("Uniswap Opt In"):
        with boa.env.prank(oeth.governor()):
            oeth.governanceRebaseOptIn(UNISWAP_POOL)
    with with_stats("Rebase After Uniswap Opt In"):
        dripper.collectAndRebase()
```

**Comment by sparrowDom** [comment]
Thanks @DanielVF you make some great points. I've applied the corrections in this commit: https://github.com/OriginProtocol/origin-dollar/pull/1961/commits/92e034da66ed568ed38f73699293fedb3ad4aea2

**Comment by sparrowDom** [comment]
@naddison36 @shahthepro this PR has a bit of a higher priority. In case you have some free cycles, could you take a look please?

File: contracts/test/token/oeth.fork-test.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  beforeEach(async () => {
+    fixture = await loadDefaultFixture();
+  });
+
+  describe("verify state", () => {
+    // These tests use a transaction to call a view function so the gas usage can be reported.
+    it("Should get total value", async () => {
+      const { oeth } = fixture;
+      const eigenLayerStrategyContract =
+        "0xa4c637e0f704745d182e4d38cab7e7485321d059";
```
Comment (naddison36) [inline contracts/test/token/oeth.fork-test.js:36]: nit: I'd add EigenLayerStrategy to the addresses file

**Review by naddison36 — APPROVED** [review]
Seems simple enough

File: contracts/test/token/oeth.fork-test.js (hunk 1 above)
Comment (sparrowDom) [reply contracts/test/token/oeth.fork-test.js:36]: This is almost the same comment I've left on one of @shahthepro's PRs but he made a great point, that when these addresses are immutable, it is safer to leave them hardcoded in the test files, so someone doesn't mistakingly change the address in the `addresses.js` and that could have unexpected consequences. [This comment](https://github.com/OriginProtocol/origin-dollar/pull/1961/files#diff-0253f024335a35ec3acd15baf09dd345175408bd79d82eda7a5ed4883c3a82aaR6-R18) addresses it.

I think Shah makes a great point, though I don't have strong conviction either way regarding this topic.

**Review by DanielVF — APPROVED** [review]

**Pull request merged**